apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -492,14 +492,24 @@ const STATUS_CODE_RESPONSE = {
   422: null,
 };
 
+/**
+ * Make HTTP fetch requests with retries and pagination handling
+ *
+ * @param input URL to visit
+ * @param opts fetch options
+ * @param incomingfetchRetryOpts options specific to fetchRetry
+ * @param previousResults Used for paginated content
+ * @returns Array of JSON responses from fetch
+ */
 export async function fetchRetry(
   input: RequestInfo | URL,
   opts?: RequestInit | undefined,
   incomingfetchRetryOpts?: {
     retryLeft?: number;
     sleepMs?: number;
   },
-) {
+  previousResults: unknown[] = [],
+): Promise<unknown[]> {
```

Comment:
I really don't like making the default response type an array. Surely we know which endpoints are going to require pagination? I'd rather `fetchRetry` focus just on retry logic, and have another wrapper (`fetchPagiated`?) that can handle pagination when we know it's expected.

Rewriting this outside of `fetchRetry` should allow us to do this iteratively instead of recursively, which I generally prefer: that avoids a DoS issue if the maximum call stack depth is exceeded.

Reply:
I think we could make guesses about if we think what we're querying might be paginated, but we don't know until we get the data if there's more than one page. For example, my names and roles calls was a single response in my test environment until we added >10 students (and then it broke). NRPS spec doesn't say anything about pagination, so we have to do custom logic to unpack and merge it. To me, returning an array of responses works for the single and paginated case in a way that the calling function can figure out to address. If the calling function only expects one page, it can use `fetchArray[0]` -- and if that's wrong, the place to fix it is in the calling function rather than `fetchRetry`.

ObRef: https://canvas.instructure.com/doc/api/file.pagination.html

I think we could unwrap this iteratively. The recursion was just for simplicity.

Reply:
My point is that things like creating a line item will _never_ result in a paginated response, so there's no need to use a paginated helper, and it could in fact be a problem for the caller if multiple things were returned. I'd rather encode expectations about data shape in the API we use to interact with LTI services, so if we _could_ reasonably expect multiple pages (as is the case with names and role service, listing line items, etc), we use a paginated version, but if we know we won't get multiple pages, don't use the paginated helper.

Reply:
I agree with the API exposing the expectations is reasonable. I'm concerned that the logic for pagination has to live in `fetchRetry` - We'd have to pass back more than `response.json()` to be able to inspect and iterate.

What if we wrap a new function `fetchRetrySingle` (or whatever) around `fetchRetry` that passes in options to not paginate and returns a single response instead of an array? I think that achieves your goals? i.e. a modified `fetchRetry` wraps around `fetchRetryPaginated` with options and post-processing?

Or we write two different functions that do fetch, but that seems worse.

Reply:
I think passing back the entire `Response` object for the single-fetch case would be reasonable. Then the implementation of `fetchPaginated` could look something like this (untested):

```ts
export async function fetchPaginated(input: RequestInfo | URL, opts?: RequestInit | undefined) {
  const responses: unknown[] = [];

  while (true) {
    const res = await fetchRetry(input, opts);
    responses.push(await res.json());

    const parsed = parseLinkHeader(res.headers.get('link')) ?? {};
    if (!('next' in parsed)) {
      return responses;
    }
  }
}
```

Seems like a really nice way to compose simple functions.

Reply:
Refactored in the latest commits.

Reply:
Thanks! Looks good.